### DISCUSSION TITLE
CURLE_OPERATION_TIMEDOUT with different situations

### DISCUSSION THREAD

**Discussion by mixapiy** [curl/curl#11119] [Q&A] [answered]
Hi. I get one CURLcode (CURLE_OPERATION_TIMEDOUT) for fundamentally different situations, for example:  

Situation 1
- Handshake was done, client disconnected on 'timeout'

Situation 2
- Handshake was not performed, client disconnected on 'connect_timeout'

I want to know if the server has received my request. How do I handle this situation correctly with a single CURLcode?

**Comment by bagder** [comment, marked as answer]
> I want to know if the server has received my request.

That's not easy to answer. When libcurl cannot complete its operation because it means a time limit, it returns the timeout error code. You can record outgoing traffic with the debug callback and then you can *guess* if sending the data means that the server has received your request or not.

**Reply by mixapiy** [reply]
Can you return a different error code?

**Reply by bagder** [reply]
> Can you return a different error code?

For what?

**Reply by mixapiy** [reply]
Unfortunately, many services do not provide idempotent queries. A more detailed error code will allow us to take faster and more accurate corrective action.

**Reply by bagder** [reply]
Perhaps, but as a client we cannot know what the server did or didn't do when the operation was not completed and the correct response was not sent back.

**Reply by mixapiy** [reply]
I agree, but you need to get a handshake back from the server before sending the data, and if it is missing during the 'connect_timeout' period, then the request is not sent, and this could be an independent error code. What do you think about this?
